netty/src/main/java/io/grpc/netty/GrpcHttp2Headers.java

```diff
@@ -0,0 +1,154 @@
+      @Nullable AsciiString path, @Nullable AsciiString method, @Nullable AsciiString scheme,
+      @Nullable AsciiString key1, @Nullable AsciiString value1, @Nullable AsciiString key2,
+      @Nullable AsciiString value2, @Nullable AsciiString key3, @Nullable AsciiString value3) {
+    normalHeaders = new AsciiString[serializedMetadata.length];
+    for (int i = 0; i < normalHeaders.length; i++) {
+      normalHeaders[i] = new AsciiString(serializedMetadata[i], false);
+    }
+    psuedoHeaders = new AsciiString[14]; // Number of extra headers * 2
+    int count = 0;
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.AUTHORITY.value(), authority, count);
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.PATH.value(), path, count);
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.METHOD.value(), method, count);
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.SCHEME.value(), scheme, count);
+    count = addPseudoHeader(key1, value1, count);
+    count = addPseudoHeader(key2, value2, count);
+    count = addPseudoHeader(key3, value3, count);
+    psuedoHeadersCount = count;
+  }
+
+  /**
+   * Server constructor.
+   */
+  GrpcHttp2Headers(byte[][] serializedMetadata, @Nullable AsciiString status,
+      @Nullable AsciiString key1, @Nullable AsciiString value1) {
+    normalHeaders = new AsciiString[serializedMetadata.length];
+    for (int i = 0; i < normalHeaders.length; i++) {
+      normalHeaders[i] = new AsciiString(serializedMetadata[i], false);
+    }
+    psuedoHeaders = new AsciiString[4]; // Number of extra headers * 2
+    int count = 0;
+    count = addPseudoHeader(Http2Headers.PseudoHeaderName.STATUS.value(), status, count);
+    count = addPseudoHeader(key1, value1, count);
+    psuedoHeadersCount = count;
+  }
+
+  private int addPseudoHeader(
+      AsciiString headerKey, @Nullable AsciiString headerValue, int currentCount) {
+    if (headerValue != null) {
+      psuedoHeaders[currentCount] = headerKey;
+      psuedoHeaders[currentCount + 1] = headerValue;
+      return currentCount + 2;
+    }
+    return currentCount;
+  }
+
+  @Override
+  public Iterator<Entry<CharSequence, CharSequence>> iterator() {
+    return new Itr();
+  }
+
+  @Override
+  public int size() {
+    return (normalHeaders.length + psuedoHeadersCount) / 2;
+  }
+
+  private class Itr implements Iterator<Entry<CharSequence, CharSequence>> {
+    private int idx;
+    private AsciiString[] currentArray;
+
+    @Override
+    public boolean hasNext() {
+      if (currentArray == null) {
+        currentArray = psuedoHeaders;
+      }
+      if (currentArray == psuedoHeaders) {
+        if (idx < psuedoHeadersCount) {
+          return true;
+        } else {
+          currentArray = normalHeaders;
+          idx = 0;
+        }
+      }
+
+      return idx < normalHeaders.length;
+    }
+
+    @Override
+    public Entry<CharSequence, CharSequence> next() {
+      final AsciiString key = currentArray[idx];
```

Comment:
This was taking advantage of DefaultHttp2HeadersEncoder, which iterates using a for-loop.

Changed to call hasNext().  This is taking advantage of the fact that preHeaders are never empty.